source/common/upstream/load_balancer_impl.cc

```diff
@@ -1103,10 +1103,38 @@ void EdfLoadBalancerBase::refresh(uint32_t priority) {
     // Cycle through hosts to achieve the intended offset behavior.
     // TODO(htuch): Consider how we can avoid biasing towards earlier hosts in the schedule across
     // refreshes for the weighted case.
-    if (!hosts.empty()) {
-      for (uint32_t i = 0; i < seed_ % hosts.size(); ++i) {
-        auto host =
-            scheduler.edf_->pickAndAdd([this](const Host& host) { return hostWeight(host); });
+    if (Runtime::runtimeFeatureEnabled("envoy.reloadable_features.edf_lb_scheduler_init_fix")) {
+      // Only apply randomization if there are at least 2 hosts. If there's one host
+      // (or no hosts), its weight doesn't matter for the load-balancer.
+      if (hosts.size() > 1) {
+        // Compute the greatest common divisor of all the weights, to normalize
+        // the sum of weights to a smaller number. This can be avoided, but will
+        // result in higher numbers and more calls to `pickAndAdd()`.
+        const uint32_t host0_weight = static_cast<uint32_t>(std::ceil(hostWeight(*hosts[0])));
+        const uint32_t host1_weight = static_cast<uint32_t>(std::ceil(hostWeight(*hosts[1])));
```

Comment:
My only concern is what will happen if the `hostWeight` return zero? Or will this happen?

Reply:
iirc, some lb will add a factor to the original weight. (Like lbs that enabled slow start)

Reply:
I'll add a lower bound to the normalizer.
@nezdolik could slow start set host weights to 0?

Reply:
Slow start code uses max(calculated_weight, min_weight_percentage) for weight, min_weight_percentage defaults to 0.1. Am not by computer now, but could paste relevant links here later today.

Reply:
This piece of code applies slow start factor in most complex logical if branch (in other logical branches original host weight is returned):
```
      return host_weight * std::max(applyAggressionFactor(time_factor, aggression),
                                    slow_start_min_weight_percent_);
```
1. If original `host_weight` is 0 computation result will be 0
2. `applyAggressionFactor(time_factor, aggression)` cannot be zero, time_factor is calculated as:
```
      auto time_factor = static_cast<double>(std::max(std::chrono::milliseconds(1).count(),
                                                      in_healthy_state_duration.count())) /
                         slow_start_window_.count();
```
to get the closest value to 0, slow_start_window_ divisor should be max value of std::chrono::milliseconds and dividend 1 millisecond, that value is still not 0 and fits into 7 bytes (double type size is 8 bytes, so should be no overflow).
3. slow_start_min_weight_percent_ can be set to 0 according to api.proto, but we have a max guard and in that case `applyAggressionFactor(time_factor, aggression)` will be selected as max, and is proven it cannot be zero in step 2.

Reply:
I relied on various online tools for floating point precision arithmetics and overflow testing like https://www.eeweb.com/tools/online-scientific-calculator/ or https://www.h-schmidt.net/FloatConverter/IEEE754.html. With 99% confidence slow start factor function cannot return 0 :) (unless original host weight is zero)